<center><h1>Teste de Hipóteses Qui-Quadrado (Chi-Squared Test)</h1></center>

# Introdução

O teste $\chi^2$ (lê-se *qui-quadrado*) é um procedimento projetado para testar a existência de relações significativas **entre duas variáveis discretas**.

## Tabela de contingência

Para aplicação do teste $\chi^2$, as informações coletadas (a partir da amostra) acerca das variáveis devem estar ​​organizadas em uma *tabela de contingência*.

Uma tabela de contingência (também conhecida como tabulação cruzada ou tabela de referência cruzada) é um tipo de tabela em um formato de matriz que exibe a distribuição de frequência das diversas combinações de valores das variáveis. Tabelas de contingência são muito utilizados em pesquisa de opinião, inteligência de negócios, engenharia e pesquisa científica. Elas fornecem uma imagem básica da inter-relação entre duas variáveis ​​e podem ajudar a encontrar interações entre elas.

Como exemplo, considere duas variáveis, ambas discretas: 

* `sexo` (com valores **Masculino** ou **Feminino**) 
* `animal de estimação` (com valores **Cachorro** ou **Gato**). 

Suponha que 100 indivíduos sejam amostrados aleatoriamente em uma população muito grande como parte de um estudo de diferenças de sexo e preferência por animal de estimação. Uma tabela de contingência pode ser criada para exibir as diversas quantidades de indivíduos na amostra: quantos são homens e preferem gatos, quantos homens preferem gatos, quantas mulheres preferem cachorros e quantas mulheres preferem gatos. Essa tabela de contingência é exibida abaixo ([fonte](https://math.stackexchange.com/questions/1233587/checking-dependency-of-variables-in-a-contingency-table)).

<center>

|               | Cachorro       | Gato       | Total |
| ---------     | -------------- | ---------- | ----- |
| **Masculino** | 42             | 10         | 52    |
| **Feminino**  | 9              | 39         | 48    |
| **Total**     | 51             | 49         | 100   |

</center>

Ainda com relação ao exemplo acima, repare que

- a coluna mais à direita apresenta os totais por linha,
- a linha mais abaixo exibe os totais por coluna,
- a célula inferior direita o total geral de indivíduos da amostra.

As informações dos totais por linha e por coluna são opcionais na representação de uma tabela de contingência. Além disso, tabelas de contingência não precisam envolver apenas variáveis binárias. Contanto que as variáveis sejam discretas (e com uma quantidade de valores dentro do aceitável), uma tabela de contingência pode ser usada. No exemplo a seguir ([fonte](http://www.sthda.com/sthda/RDoc/data/housetasks.txt)), as variáveis apresentam quatro e treze valores, respectivamente. Note também que não são fornecidos os totais nas margens da tabela.

<center>

| Task        | Wife | Alternating | Husband | Jointly |
|-------------|------|-------------|---------|---------|
| Laundry     | 156  | 14          | 2       | 4       |
| Main_meal   | 124  | 20          | 5       | 4       |
| Dinner      | 77   | 11          | 7       | 13      |
| Breakfast   | 82   | 36          | 15      | 7       |
| Tidying     | 53   | 11          | 1       | 57      |
| Dishes      | 32   | 24          | 4       | 53      |
| Shopping    | 33   | 23          | 9       | 55      |
| Official    | 12   | 46          | 23      | 15      |
| Driving     | 10   | 51          | 75      | 3       |
| Finances    | 13   | 13          | 21      | 66      |
| Insurance   | 8    | 1           | 53      | 77      |
| Repairs     | 0    | 3           | 160     | 2       |
| Holidays    | 0    | 1           | 6       | 153     |

</center>


## Independência estatística

A independência estatística é um conceito importante na Teoria das Probabilidades. Considere dois eventos $A$ e $B$. Se o conhecimento da ocorrência do evento $A$ não muda a probabilidade de ocorrência do evento $B$, então escrevemos:

$$
\Pr(B|A) = \Pr(B)
$$

Sob a condição acima, a probabilidade conjunta dos eventos $A$ e $B$ é igual ao produto das probabilidades dos eventos $A$ e $B$: $\Pr(A,B) = \Pr(A) \times \Pr(B)$. Esses eventos são então considerados **estatisticamente independentes**.

A independência estatística não deve ser confundida com outra propriedade que pode existir entre eventos, a exclusividade mútua. Se a probabilidade conjunta de dois eventos $A$ e $B$ for zero, esses eventos são ditos mutuamente exclusivos ou disjuntos. O seguinte destaca as condições sob as quais dois eventos são estatisticamente independentes e mutuamente exclusivos:

- $\Pr(A,B) = \Pr(A) \times \Pr(B) \iff \text{eventos independentes}$
- $\Pr(A,B) = 0 \iff \text{eventos mutuamente exclusivos}$

No contexto da Inferência Estatística, o conceito de **evento** está relacionado aos valores que uma variável aleatória assume. Além disso, o conceito de independência estatística entre duas variáveis aleatórias pode ser usado como um indicador de presença ou ausência de dependência entre elas. Em particular, se duas variáveis são independentes, as distribuições percentuais na tabela de contingência são idênticas.

> Se há dependências entre duas variáveis, dizemos que essas variáveis estão **associadas**.

Como exemplo, considere novamente a tabela de contingência usada em exemplo anterior e reapresentada abaixo. 

<center>

|               | Cachorro       | Gato       | Total |
| ---------     | -------------- | ---------- | ----- |
| **Masculino** | 42             | 10         | 52    |
| **Feminino**  | 9              | 39         | 48    |
| **Total**     | 51             | 49         | 100   |

</center>

Nesse contexto, há duas variáveis aleatórias binárias: 
* **Sexo** 
* **Preferência** (por animal de estimação preferido)

Os dados nessa tabela parecem indicar que as variáveis **não** são estatisticamente independentes, i.e., que há associação entre elas. Para ententer porque, repare que a maioria das mulheres prefere gatos (39 contra 9), enquanto que a maioria dos homens prefere cachorros (42 contra 10). Se essas variáveis fossem efetivamente independentes, deveríamos esperar encontrar preferências aproximadamente iguais por gatos e cachorros, tanto no grupo de mulheres quanto no de homens.

Visto que os dados apresentados na tabela acima são correspondentes a uma amostra da população, surge a seguinte questão:

> O enviesamento observado é obra do acaso (e portanto não reflete a distribuição da população), ou é consequência de as variáveis serem estatisticamente dependentes?

A questão acima pode ser investigada por meio do teste $\chi^2$.

## Distribuição Qui-Quadrado

A estatística usada no teste $\chi^2$ segue a distribuição homônima, a [distribuição $\chi^2$](https://en.wikipedia.org/wiki/Chi-squared_distribution). A distribuição qui-quadrado (𝜒²) é uma distribuição contínua assimétrica. A variável aleatória $X \sim \chi^2_k$ com $k$ graus de liberdade é a soma dos quadrados de $k$ variáveis normais padrão independentes:

$$
X = \sum_{i=1}^{k} Z_i^2, \quad Z_i \sim N(0,1)
$$

A figura a seguir ([fonte](https://en.wikipedia.org/wiki/Chi-squared_distribution)) apresenta os gráficos de diversas funções de densidade que seguem a distribuição $\chi^2$. Essa distribuição possui um parâmetro, $k$, a quantidade de graus de liberdade.

<p align="center">
  <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/35/Chi-square_pdf.svg/1024px-Chi-square_pdf.svg.png" width="600"/>
</p>

A biblioteca scipy disponibiliza o pacote de funções [chi2](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html), que permite manipular variáveis aleatórias que seguem a distribuição chi-quadrado.

Como exemplo, suponha uma variável aleatória que siga a distribuição qui-quadrado com 9 graus de liberdade. Se quisermos saber que valor dessa variável delimita 95% da distribuição, podemos usar a função `ppf`, conforme o bloco de código abaixo.

In [34]:
from scipy.stats import chi2

print(chi2.ppf(.95, df=2))
print(chi2.cdf(5.9914, df=2))


5.991464547107979
0.9499983862962607


## Tipos de Testes Qui-Quadrado

### 1. Teste de Aderência (Goodness-of-fit)

Verifica se uma distribuição observada se ajusta a uma distribuição teórica esperada. Neste tipo de teste, testamos se uma determinada amostra é proveniente de uma suposta distribuição de probabilidade;

**Declaração das hipóteses:**
* $H_0$: A distribuição observada segue a distribuição teórica.
* $H_a$: A distribuição observada não segue a distribuição teórica.

### 2. Teste de Independência

Verifica se há associação (i.e., dependência estatística) entre duas variáveis categóricas. Neste teste, temos interesse em verificar se as variáveis com distribuições nas marginais são independentes ou não.


**Declaração das hipóteses:**
* $H_0$: As variáveis são independentes.
* $H_a$: As variáveis são associadas.


Neste tipo de teste $\chi^2$, as hipóteses devem ser declaradas da seguinte forma:
- A hipótese nula ($H_0$) deve afirmar que não existe associação entre as duas variáveis.
- A hipótese de pesquisa ($H_a$) declara que as duas variáveis ​estão associadas na população.

### 3. Teste de Homogeneidade

Verifica se **duas ou mais populações** seguem a **mesma distribuição** de uma variável categórica. Em outras palavras, testamos se uma determinada variável se distribui da mesma forma em várias populações de interesse.

- $H_0$: Todas as probabilidades numa mesma colunas são iguais ($p_{1j} = \ldots = p_{rj} = p_j$, para todo $j$).
- $H_a$: Pelo menos uma das probabilidades é diferente de $p_{ij}$ é diferente de $p_j$, para algum $j$

---

# Estatística qui-quadrado

A estatística qui-quadrado mede o desvio entre o que foi observado nos dados e o que seria esperado caso não houvesse relação entre as variáveis (sob $H_0$). Essa estatística é calculada por meio da consulta aos valores contidos em uma tabela de contingência.

$$
\chi^2 = \sum_{i,j} \frac{(O_{ij} - E_{ij})^2}{E_{ij}}
$$

* $O_{ij}$: **Frequência observada**. Frequências observadas (*observed frequencies*) são valores efetivamente observados na tabela de contingência (ou derivados dela).
* $E_{ij}$: **Frequência esperada sob $H_0$**. Frequências esperadas (*expected frequencies*) são valores que seriam esperados se as duas variáveis fossem estatisticamente independentes. Calculamos uma frequência observada para cada par de valores das variáveis.

> A estatística qui-quadrado é uma soma de desvios normalizados: se o observado é muito diferente do esperado, o numerador é grande, e a estatística assume valores grandes. 

Para computar a estatística $\chi^2$, é necessário computar as frequências esperadas para cada célula $(i,j)$ da tabela de contingência. Para isso, basta multiplicar os totais de linha e de coluna para essa célula e dividir o produto resultante pelo número total de casos na tabela (representado por $N$ na equação abaixo).

$$
E_{ij} = \frac{{\text{total da linha $i$} \times \text{total da coluna $j$}}}{N}
$$

Como exemplo, retornemos à tabela de contingência usada em exemplo anterior e reapresentada abaixo.

<center>

|               | Cachorro       | Gato       | Total |
| ---------     | -------------- | ---------- | ----- |
| **Masculino** | 42             | 10         | 52    |
| **Feminino**  | 9              | 39         | 48    |
| **Total**     | 51             | 49         | 100   |

</center>

Se não houver associação entre as variáveis **Sexo** e **Preferência**, as frequências esperadas são calculadas como:

- $E_{11} = f_e$(Masculino, Cachorro) = $\frac{51 \times 52}{100} = 26.52$

- $E_{12} = f_e$(Masculino, Gato) = $\frac{49 \times 52}{100} = 25.48$

- $E_{21} = f_e$(Feminino, Cachorro) = $\frac{51 \times 48}{100} = 24.48$

- $E_{22} = f_e$(Feminino, Gato) = $\frac{49 \times 48}{100} = 23.52$

Podemos organizar essas frequências esperadas em outra tabela de contingência:
<center>

| Sexo      | Cachorro | Gato  |
| --------- | ---------| ------|
| Masculino | 26.52    | 25.48 |
| Feminino  | 24.48    | 23.52 |

</center>

De posse das frequências esperadas e observadas, é possível computar o valor da estatística:

$$
\chi^2 = \frac{(26.52-42)^2}{26.52} + \frac{(25.48-10)^2}{25.48} + \frac{(24.48-9)^2}{24.48} + \frac{(23.52-39)^2}{23.52} \approx 38.41
$$

A célula de código abaixo realiza o mesmo cálculo, para fins de conferência.

In [35]:
import numpy as np

exp_11 = 51*52/100
exp_12 = 49*52/100
exp_21 = 51*48/100
exp_22 = 49*48/100
print(exp_11)
print(exp_12)
print(exp_21)
print(exp_22)

qui_square = np.square(exp_11 - 42)/exp_11 + np.square(exp_12 - 10)/exp_12 + np.square(exp_21 - 9)/exp_21 + np.square(exp_22 - 39)/exp_22
print(qui_square)

26.52
25.48
24.48
23.52
38.41767476221258


# Passos do procedimento

1. Declarar as hipóteses nula e alternativa
2. Montar a tabela de contingência
3. Computar a estatística de teste
4. Calcular o valor $p$ correspondente
5. Tomar uma decisão e interpretar o resultado

# Aplicação do teste

No contexto do teste $\chi^2$, a distribuição $\chi^2$
indica a probabilidade de obter um determinado valor, presumindo que não existe relação (na população) entre as duas variáveis. Ainda no contexto desse teste, o valor do parâmetro $k$ é calculado por meio da expressão

$$
k = (r – 1) \times (c – 1)
$$

onde
- $r$ = quantidade de linhas da tabela, e
- $c$ = quantidade de colunas da tabela.

Seguindo com o exemplo anterior, o parâmetro $k$ pode ser calcular conforme a seguir:

$$
k = (2-1)(2-1) = 1
$$

De posse dos valores da estatística ($\chi^2 = 38.41$) e do parâmetro ($k=1$), podemos computar o valor corresponde de $p$.

Considere que foi definido um nível de significância igual a 5% para o estudo estatístico. Usando a função [scipy.stats.chi2.ppf](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2.html#scipy.stats.chi2), concluímos que o valor $3.84$, conhecido como valor crítico, é o que separa 5% da área da curva à direita. Por outro lado, o valor $\chi^2 = 38.41$ é muito maior do que $3.84$. De fato, o valor obtido para a estatística está relacionado um valor $p$ virtualmente igual a $0$. Então parece muito claro que as variáveis ​​categóricas são **dependentes** na população da qual a amostra de 100 observações foi escolhida aleatoriamente.

In [36]:
from scipy.stats import chi2
x = 38.41
k = 1
print('valor crítico: %.2f' % chi2.ppf(.95, k))
print('p: %f' % (1 - chi2.cdf(x,k)))

valor crítico: 3.84
p: 0.000000


Uma forma mais direta de aplicar o teste qui-quadrado é usar a função [`scipy.stats.chisquare`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html). 

> `chisquare(f_obs, f_exp=None, ddof=0, axis=0, *, sum_check=True)`

Veja o exemplo abaixo.

In [37]:
from scipy.stats import chisquare
chisquare(f_obs = [42, 10, 9, 39],
          f_exp = [26.52, 25.48, 24.48, 23.52], 
          ddof = 1)

Power_divergenceResult(statistic=38.41767476221258, pvalue=4.54682167469138e-09)

# Condições de aplicabilidade

Diferente dos testes estudados até aqui ($z$-teste, $t$-teste, ANOVA), o teste $\chi^2$ não requer suposições sobre a distribuição da população da qual a amostra é retirada. Ou seja, o $\chi^2$ é um teste estatístico não-paramétrico.

No entanto, assim como outras técnicas inferenciais, esse teste possui algumas consições de aplicabilidade:

* As variáveis do estudo devem ser categóricas.

* As observações devem ser independentes (i.e., presume que foi realizada amostragem aleatória.). Cada resposta deve vir de uma unidade estatística **independente** das demais.

* A frequência esperada de cada célula da tabela de contingência deve ser razoavelmente grande. Como regra prática, costuma-se usar o seguinte:
  * Nenhuma célula com frequência esperada < 1.
  * No máximo 20% das células com frequência esperada < 5.

Avaliar as **condições de aplicabilidade** é essencial antes de confiar nos resultados do teste qui-quadrado.



# Exemplos

Nos exemplos a seguir, considere $\alpha = 0.05$.

## Exemplo

Vamos aplicar o teste aos dados dos casais, com suas tarefas domésticas e definição do(s) executor(es).
<center>

| Task        | Wife | Alternating | Husband | Jointly |
|-------------|------|-------------|---------|---------|
| Laundry     | 156  | 14          | 2       | 4       |
| Main_meal   | 124  | 20          | 5       | 4       |
| Dinner      | 77   | 11          | 7       | 13      |
| Breakfast   | 82   | 36          | 15      | 7       |
| Tidying     | 53   | 11          | 1       | 57      |
| Dishes      | 32   | 24          | 4       | 53      |
| Shopping    | 33   | 23          | 9       | 55      |
| Official    | 12   | 46          | 23      | 15      |
| Driving     | 10   | 51          | 75      | 3       |
| Finances    | 13   | 13          | 21      | 66      |
| Insurance   | 8    | 1           | 53      | 77      |
| Repairs     | 0    | 3           | 160     | 2       |
| Holidays    | 0    | 1           | 6       | 153     |

</center>



In [48]:
import pandas as pd
from scipy.stats import chi2_contingency

# Criar o DataFrame com os dados da tabela
data = {
    "Wife": [156, 124, 77, 82, 53, 32, 33, 12, 10, 13, 8, 0, 0],
    "Alternating": [14, 20, 11, 36, 11, 24, 23, 46, 51, 13, 1, 3, 1],
    "Husband": [2, 5, 7, 15, 1, 4, 9, 23, 75, 21, 53, 160, 6],
    "Jointly": [4, 4, 13, 7, 57, 53, 55, 15, 3, 66, 77, 2, 153]
}

tasks = [
    "Laundry", "Main_meal", "Dinner", "Breakfast", "Tidying", "Dishes", "Shopping",
    "Official", "Driving", "Finances", "Insurance", "Repairs", "Holidays"
]

df = pd.DataFrame(data, index=tasks)

# Aplicar o teste qui-quadrado
chi2_stat, p_val, dof, expected = chi2_contingency(df)

chi2_stat, p_val, dof, pd.DataFrame(expected, index=tasks, columns=df.columns)

(1944.4561959955277,
 0.0,
 36,
                 Wife  Alternating    Husband    Jointly
 Laundry    60.550459    25.633028  38.449541  51.366972
 Main_meal  52.637615    22.283257  33.424885  44.654243
 Dinner     37.155963    15.729358  23.594037  31.520642
 Breakfast  48.165138    20.389908  30.584862  40.860092
 Tidying    41.972477    17.768349  26.652523  35.606651
 Dishes     38.876147    16.457569  24.686353  32.979931
 Shopping   41.284404    17.477064  26.215596  35.022936
 Official   33.027523    13.981651  20.972477  28.018349
 Driving    47.821101    20.244266  30.366399  40.568234
 Finances   38.876147    16.457569  24.686353  32.979931
 Insurance  47.821101    20.244266  30.366399  40.568234
 Repairs    56.766055    24.030963  36.046445  48.156537
 Holidays   55.045872    23.302752  34.954128  46.697248)

**Resultado do Teste**

- Estatística $\chi^2$: 1944.46

- Valor-p: ~ 0.0 (menor que qualquer nível de significância usual)

- Graus de liberdade (GL): 36

**Conclusão**: Rejeitamos fortemente a hipótese nula. Há evidência estatística de que a distribuição de quem realiza a tarefa depende do tipo de tarefa — ou seja, há associação entre tarefa e executor.

---

## Exemplo 0 - Condições de Aplicabilidade

Vamos verificar as **condições de aplicabilidade** para os dados dos casais com suas tarefas domésticas e atribuições.

### 1. **Variáveis categóricas?**

✅ **Verifica?** → **Sim**

Neste caso, temos duas variáveis categóricas:
- `Tarefa` (com valores `Laundry`, `Main_meail`, ..., `Holydays`)
- `Atribuição` (com valores `Wife`, `Alternating`, `Husband`, `Jointly`)

### 2. **Independência das observações**

✅ **Verifica?** → **Sim**, presumivelmente cada linha corresponde a tarefas distintas e cada contagem é independente.

**Suposição razoável** se os dados foram coletados por pessoa/tarefa.

### 3. **Frequências esperadas razoavelmente grandes**

Vamos verificar isso diretamente por meu de um código em Python:

In [38]:
import numpy as np

# Flatten a matriz de frequências esperadas
flat_expected = expected.flatten()
less_than_5 = np.sum(flat_expected < 5)
less_than_1 = np.sum(flat_expected < 1)
total_cells = flat_expected.size

percent_below_5 = 100 * less_than_5 / total_cells
less_than_1, less_than_5, percent_below_5

(0, 0, 0.0)

Verificação das Condições Numéricas:

* **Células com frequência esperada < 1**: `0`
* **Células com frequência esperada < 5**: `0`
* **Percentual de células com frequência esperada < 5**: `0.0%`

✅ **Verifica?** → **Sim**

### Conclusão:

As **condições de aplicabilidade estão plenamente satisfeitas**:

* As observações são presumivelmente independentes.
* As frequências esperadas são todas suficientemente grandes.

Portanto, **o uso do teste qui-quadrado é válido e apropriado** para este conjunto de dados.

---

## Exemplo 1: Teste de Independência

O conjunto de dados survey contém dados sobre hábitos de fumo e de atividades físicas de estudantes, dentre outras informações.

- hábitos de tabagismo (`Smoke`)
- nível de prática de exercícios físicos (`Exer`)

Seja usar o teste $\chi^2$ para verificar, nível de significância de 5%, se as variáveis são independentes.

In [39]:
import pandas as pd
df_survey = pd.read_csv('../data/survey.csv')
df_survey[['Smoke', 'Exer']].head()

Smoke  Exer
0  Never  Some
1  Regul   NaN
2  Occas   NaN
3  Never   NaN
4  Never  Some

O pacote Pandas fornece a função [crosstab](https://pandas.pydata.org/docs/reference/api/pandas.crosstab.html), usada para montar a tabela de contingência para a aplicação posterior do teste $\chi^2$. Veja a célula de código a seguir.

In [40]:
tbl = pd.crosstab(df_survey.Smoke, df_survey.Exer)
tbl

Exer   Freq  Some
Smoke            
Heavy     7     3
Never    87    84
Occas    12     4
Regul     9     7

Por inspeção da tabela acima, percebemos que a variável `Exer` contém três valores, e a variável `Smoke` contém quatro valores.

De posse da tabela de contingência, podemos usar a função [scipy.stats.chi2_contingency](scipy.stats.chi2_contingency) para aplicar diretamente o teste $\chi^2$.

In [41]:
from scipy.stats import chi2_contingency
chi2_contingency(tbl)

Chi2ContingencyResult(statistic=4.574966550693501, pvalue=0.20570012488442274, dof=3, expected_freq=array([[ 5.39906103,  4.60093897],
       [92.32394366, 78.67605634],
       [ 8.63849765,  7.36150235],
       [ 8.63849765,  7.36150235]]))

Ao ser invocada, a função chi2_contingency retorna uma tupla de componentes conforme descrição abaixo:

1. o valor da estatística qui-quadrado ($4.57$) ;
2. o $p$-valor ($0.2057$);
3. a quantidade de graus de liberdade (3);
4. as frequências esperadas, na forma de uma matriz.

Porque o valor de $p = 0.2057$ é maior do que o nível de significância de $\alpha=0.05$, não temos evidência para rejeitar a hipótese nula de que o hábito de fumar seja independente do nível de exercício dos alunos.

No resultado acima, repare que nem todas as frequências esperadas são maiores que 5. Em geral, os resultados do teste $\chi^2$ são confiáveis apenas quando todas as frequências esperadas são maiores ou iguais a 5. Para contornar esse problema, podemos combinar os valores 'None' e 'Some' da variável `Exer` em um único valor, conforme a célula de código abaixo.

In [42]:
df_survey.replace({'Exer': {'None': 'None_or_Some', 'Some': 'None_or_Some'}}, inplace=True)
df_survey[['Smoke', 'Exer']].head()

Smoke          Exer
0  Never  None_or_Some
1  Regul           NaN
2  Occas           NaN
3  Never           NaN
4  Never  None_or_Some

In [43]:
tbl = pd.crosstab(df_survey.Smoke, df_survey.Exer)
tbl

Exer   Freq  None_or_Some
Smoke                    
Heavy     7             3
Never    87            84
Occas    12             4
Regul     9             7

In [44]:
from scipy.stats import chi2_contingency
chi2_contingency(tbl)

Chi2ContingencyResult(statistic=4.574966550693501, pvalue=0.20570012488442274, dof=3, expected_freq=array([[ 5.39906103,  4.60093897],
       [92.32394366, 78.67605634],
       [ 8.63849765,  7.36150235],
       [ 8.63849765,  7.36150235]]))

Perceba que agora todos os valores das frequências esperadas estão acima de 5, fazendo com que o resultado de aplicação do $\chi^2$ dessa vez seja confiável.

---

## Exemplo 2: Teste de Aderência

Suponha que uma empresa afirma que 25% dos seus produtos são A, 25% B, 25% C e 25% D. Observamos a seguinte distribuição após uma amostra:

<center>

| Categoria | Frequência Observada |
| --------- | -------------------- |
| A         | 20                   |
| B         | 18                   |
| C         | 22                   |
| D         | 40                   |

</center>

Com esses dados, queremos verificar a alegação da empresa de que produz uniformemente produtos dos quatro tipos.

In [45]:
from scipy.stats import chisquare

observed = [20, 18, 22, 40]
expected = [25, 25, 25, 25]  # frequências esperadas uniformes

chi2_stat, p_val = chisquare(f_obs=observed, f_exp=expected)

print(f'Estatística qui-quadrado: {chi2_stat:.4f}')
print(f'Valor-p: {p_val:.4f}')

Estatística qui-quadrado: 12.3200
Valor-p: 0.0064


O valor-p é menor do que $\alpha = 0.05$. Sendo assim, rejeitamos $H_0$: os dados **não seguem** uma distribuição uniforme.

---

## Exemplo 3: Teste de Independência com Tabela de Contingência

São coletados dados de prefêrencia por um novo produto, conforme a tabela a seguir.

<center>

|           | Gosta | Não gosta |
| --------- | ----- | --------- |
| Masculino | 30    | 20        |
| Feminino  | 20    | 30        |

</center>

Será que **gênero** está associado ao **gosto pelo novo produto**?

In [46]:
import numpy as np
from scipy.stats import chi2_contingency

# Criar a tabela
table = np.array([[30, 20],
                  [20, 30]])

chi2_stat, p_val, dof, expected = chi2_contingency(table)

print(f'Estatística qui-quadrado: {chi2_stat:.4f}')
print(f'Valor-p: {p_val:.4f}')
print(f'GL: {dof}')
print('Frequências esperadas sob H0:')
print(expected)

Estatística qui-quadrado: 3.2400
Valor-p: 0.0719
GL: 1
Frequências esperadas sob H0:
[[25. 25.]
 [25. 25.]]


O valor-p não é menor do que $\alpha = 0.05$. Sendo assim, falhamos em rejeitar $H_0$: não há evidência de associação entre gênero e preferência.

---

## Exemplo 4: Teste de Homogeneidade 

Uma rede de supermercados deseja saber se **o padrão de preferência por tipos de pagamento** (dinheiro, cartão, PIX) é o **mesmo em três lojas diferentes**.

São coletados dados de clientes dessas três lojas. Dados observados:

<center>

| Forma de Pagamento | Loja A | Loja B | Loja C |
| ------------------ | ------ | ------ | ------ |
| Dinheiro           | 30     | 25     | 20     |
| Cartão             | 45     | 40     | 35     |
| PIX                | 25     | 35     | 45     |

</center>

Hipóteses do Teste:

* **$H_0$**: As distribuições de forma de pagamento são iguais nas três lojas (homogeneidade).
* **$H_1$**: Pelo menos uma loja tem uma distribuição diferente (não homogênea).

In [47]:
import numpy as np
from scipy.stats import chi2_contingency

# Tabela de frequências observadas
table = np.array([
    [30, 25, 20],  # Dinheiro
    [45, 40, 35],  # Cartão
    [25, 35, 45]   # PIX
])

# Aplicar o teste qui-quadrado
chi2_stat, p_val, dof, expected = chi2_contingency(table)

print(f'Estatística qui-quadrado: {chi2_stat:.4f}')
print(f'Valor-p: {p_val:.4f}')
print(f'Graus de liberdade: {dof}')
print('Frequências esperadas sob H0:')
print(expected)

Estatística qui-quadrado: 8.9643
Valor-p: 0.0620
Graus de liberdade: 4
Frequências esperadas sob H0:
[[25. 25. 25.]
 [40. 40. 40.]
 [35. 35. 35.]]


Interpretação

* **Se o valor-p < 0.05**, rejeitamos $H_0$: há **evidência de que as distribuições diferem entre as lojas**.
* **Se o valor-p ≥ 0.05**, não rejeitamos $H_0$: não há evidência suficiente para afirmar que as lojas têm distribuições diferentes.

Repare que o teste de homogeneidade é estruturalmente idêntico ao teste de independência (mesma estatística), mas a **interpretação muda**:

* **Teste de independência**: duas variáveis em uma única amostra.
* **Teste de homogeneidade**: uma única variável categórica observada em **grupos diferentes**.

---

## Exercício Proposto

Considere os dados:

<center>

| Cor do Carro | Acidente | Sem Acidente |
| ------------ | -------- | ------------ |
| Vermelho     | 15       | 35           |
| Preto        | 30       | 45           |
| Branco       | 25       | 50           |

</center>

1. Formule as hipóteses para testar se a cor do carro está associada à ocorrência de acidentes.
2. Aplique o teste de independência em Python.
3. Interprete os resultados.